[llvm][mlir][flang][OpenMP] Emit __atomic_load and __atomic_compare_exchange libcalls for complex types in atomic update

NimishMishra opened the pull request:
This patch adds functionality to emit relevant libcalls in case atomicrmw instruction can not be emitted (for instance, in case of complex types). The IRBuilder is modified to directly emit __atomic_load and __atomic_compare_exchange libcalls. The added functions follow a similar codegen path as Clang, so that LLVM Flang generates almost similar IR as Clang.

Fixes https://github.com/llvm/llvm-project/issues/83760 and https://github.com/llvm/llvm-project/issues/75138

Co-authored-by: Michael Kruse <llvm-project@meinersbur.de>

NimishMishra commented:
Hello all,

   For this particular patch, my main approach was to leverage AtomicExpandPass in order to automatically expand atomic load and atomic compare exchange into relevant libcalls (and as a result, refactor such code from Clang too). However, I was not able to get the pass run on desired function housing these instructions. Hence, I extracted out the necessary functionality and created this PR. Please do suggest if there is a way to improve upon this.

For now, this patch fixes the issue of complex types in atomic. For instance, given the following source:

```
program main
      complex*8 ia, ib
      ia = (2, 2)
      
      !$omp parallel num_threads(50)
      !$omp atomic update
        ia = ia + (1, 1)
      !$omp end atomic  
     !$omp end parallel
      print *, ia
end program main
```

The final value printed with `flang-new -fopenmp <test>.f90 -latomic` is `(52., 52.)`.

kiranchandramohan commented:
I was looking through a C example (given below),
```
#include <complex.h>
#include <stdio.h>
#include <tgmath.h>

float complex sub(float complex *z1, float complex *z2)
{
    #pragma omp atomic update
    *z1 = *z1 + *z2;

    return *z1;
}
```

and the LLVM IR generated by Clang looks a bit different (as given below)
```
  call void @__atomic_load(i64 noundef 8, ptr noundef %0, ptr noundef %atomic-temp, i32 noundef 0)
  br label %atomic_cont

atomic_cont:                                      ; preds = %atomic_cont, %entry
  %atomic-temp.realp = getelementptr inbounds { float, float }, ptr %atomic-temp, i32 0, i32 0
  %atomic-temp.real = load float, ptr %atomic-temp.realp, align 4
  %atomic-temp.imagp = getelementptr inbounds { float, float }, ptr %atomic-temp, i32 0, i32 1
  %atomic-temp.imag = load float, ptr %atomic-temp.imagp, align 4
  %add.r = fadd float %atomic-temp.real, %.real
  %add.i = fadd float %atomic-temp.imag, %.imag
  %atomic-temp1.realp = getelementptr inbounds { float, float }, ptr %atomic-temp1, i32 0, i32 0
  %atomic-temp1.imagp = getelementptr inbounds { float, float }, ptr %atomic-temp1, i32 0, i32 1
  store float %add.r, ptr %atomic-temp1.realp, align 4
  store float %add.i, ptr %atomic-temp1.imagp, align 4
  %call = call i1 @__atomic_compare_exchange(i64 noundef 8, ptr noundef %0, ptr noundef %atomic-temp, ptr noundef %atomic-temp1, i32 noundef 0, i32 noundef 0)
  br i1 %call, label %atomic_exit, label %atomic_cont
  ```

The code that gets called is `AtomicInfo::EmitAtomicUpdateLibcal` in `clang/lib/CodeGen/CGAtomic.cpp`. Is this the code that you used? Did you face any issue using this code? 
https://github.com/llvm/llvm-project/blob/33bf08ec36efc2fc2df8217eddd751cef9bc6be6/clang/lib/CodeGen/CGAtomic.cpp#L1764

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1640,8 +1640,7 @@ convertOmpAtomicUpdate(omp::AtomicUpdateOp &opInst,
   if (!isRegionArgUsed)
-    return opInst.emitError("no atomic update operation with region argument"
-                            " as operand found inside atomic.update region");
+    shouldEmitLibCall = true;
```
The code here has changed (as you know). 

The code in Clang seems to do the following for this boolean,
```
      UseLibcall = !C.getTargetInfo().hasBuiltinAtomic(
          AtomicSizeInBits, C.toBits(lvalue.getAlignment()));
 ```
 
 We should also probably do something similar and probably delegate this to the OpenMPIRBuilder to decide. If that is not possible then an alternative would be to look at the block argument and if it is a struct then use LibCall.

NimishMishra replied:
Okay thanks. I'll make this change in the next update

Meinersbur commented:
What happened to the idea to refactor the language-independent part of CGAtomic into LLVMFrontend to avoid duplicating the functionality?

NimishMishra commented:
Thanks @kiranchandramohan and @Meinersbur for the comments.

I did try making the IR emitted here closer to Clang, as the first option. From what I understand, there are differences in the way `__atomic_compare_exchange` is emitted by Clang or when it is expanded in AtomicExpandPass. One main difference is emission of `llvm.lifetime` calls. Then there are several `store` and `extract`.

I sent across this PR for review to request for suggestions on this only. Is it okay if Clang is also made dependent on the OpenMPIRBuilder for emitting `__atomic_compare_exchange`, when the two IRs are different? Moreover, even though I was able to emit these calls from the IRBuilder, I am not 100% sure why Clang and AtomicExpandPass choose to emit different IRs. Which one is the better one?

NimishMishra commented:
> The code that gets called is `AtomicInfo::EmitAtomicUpdateLibcal` in `clang/lib/CodeGen/CGAtomic.cpp`. Is this the code that you used? Did you face any issue using this code?

Hi @kiranchandramohan. The main issue I faced was determining which of the two IRs (Clang codegen vs AtomicExpandPass) was the one we should try. I could not determine the same, so thought to send this initial PR once, and seek clarifications on the same.

Meinersbur commented:
> I sent across this PR for review to request for suggestions on this only. Is it okay if Clang is also made dependent on the OpenMPIRBuilder for emitting `__atomic_compare_exchange`, when the two IRs are different? 

Clang shouldn't use OpenMPIRBuilder for non-OpenMP stuff. The approach would move [`AtomicInfo`](https://github.com/llvm/llvm-project/blob/9f449c34278191193f2f2cbc96c333548ad20238/clang/lib/CodeGen/CGAtomic.cpp#L30) (that is, the non-Clang specific parts of it) to LLVMFrontend which is then used by `OpenMPIRBuilder`.


> The main issue I faced was determining which of the two IRs (Clang codegen vs AtomicExpandPass) was the one we should try. I could not determine the same, so thought to send this initial PR once, and seek clarifications on the same.

The answer is both. Comments in CGAtomic justify why part is lowered there, part in `AtomicExpandPass`. 
Regarding lifetime markers, orienting towards what Clang does is probably less risky. `AtomicExpandPass` is used only at the end of the optimization pipeline, passes in-between may already have transformed the lifetime markers.
Take [IR Best Practices](https://llvm.org/docs/Frontend/PerformanceTips.html#ir-best-practices) into account:

> [...] it’s worth noting that the most mature frontend for LLVM is Clang. As a result, the further your IR gets from what Clang might emit, the less likely it is to be effectively optimized.

unless there is a reason not to.

NimishMishra commented:
Thanks @Meinersbur for the clarification. I'll work on these changes then. I think I'll begin from studying some existing functionality wherein Clang interfaces with LLVM Frontend, and then follow similar steps to refactor atomic libcall emission code into the LLVM Frontend. This should let Clang emit libcalls just as it does now. Finally, we can allow the IRBuilder to use this code in the LLVM Frontend to let Flang also depend upon the same code.

Could you suggest some existing functionality where Clang interfaces with LLVM frontend? I'll explore anyway; but if you have something in mind, please do share.

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -6033,6 +6034,52 @@ std::pair<Value *, Value *> OpenMPIRBuilder::emitAtomicUpdate(
       Res.second = Res.first;
     else
       Res.second = emitRMWOpAsInstruction(Res.first, Expr, RMWOp);
+  } else if (RMWOp == llvm::AtomicRMWInst::BinOp::BAD_BINOP) {
```
`BAD_BINOP`? Doesn't this implement `BinOp::Xchg`?

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -6033,6 +6034,52 @@ std::pair<Value *, Value *> OpenMPIRBuilder::emitAtomicUpdate(
+    Builder.CreateStore(Upd, NewAtomicAddr);
+    AtomicOrdering Failure =
+        llvm::AtomicCmpXchgInst::getStrongestFailureOrdering(AO);
+    auto Result = atomicInfo.EmitAtomicCompareExchangeLibcall(
```
```suggestion
    auto Result = atomicInfo.EmitAtomicCompareExchange(
```
The decision of whether to to use libcall should be made by AtomicInfo itself.

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -6033,6 +6034,52 @@ std::pair<Value *, Value *> OpenMPIRBuilder::emitAtomicUpdate(
+    unsigned LoadSize =
+        LoadDL.getTypeStoreSize(OldVal->getPointerOperand()->getType());
+    OpenMPIRBuilder::AtomicInfo atomicInfo(&Builder, XElemTy, LoadSize * 8,
+                                           LoadSize * 8, OldVal->getAlign(),
+                                           OldVal->getAlign(), true, X);
+    auto AtomicLoadRes = atomicInfo.EmitAtomicLoadLibcall(AO);
+    BasicBlock *CurBB = Builder.GetInsertBlock();
+    Instruction *CurBBTI = CurBB->getTerminator();
+    CurBBTI = CurBBTI ? CurBBTI : Builder.CreateUnreachable();
+    BasicBlock *ExitBB =
+        CurBB->splitBasicBlock(CurBBTI, X->getName() + ".atomic.exit");
+    BasicBlock *ContBB = CurBB->splitBasicBlock(CurBB->getTerminator(),
+                                                X->getName() + ".atomic.cont");
+    ContBB->getTerminator()->eraseFromParent();
+    Builder.restoreIP(AllocaIP);
+    AllocaInst *NewAtomicAddr = Builder.CreateAlloca(XElemTy);
+    NewAtomicAddr->setName(X->getName() + "x.new.val");
+    Builder.SetInsertPoint(ContBB);
+    llvm::PHINode *PHI = Builder.CreatePHI(OldVal->getType(), 2);
+    PHI->addIncoming(AtomicLoadRes.first, CurBB);
+    Value *OldExprVal = PHI;
+    Value *Upd = UpdateOp(OldExprVal, Builder);
+    Builder.CreateStore(Upd, NewAtomicAddr);
+    AtomicOrdering Failure =
+        llvm::AtomicCmpXchgInst::getStrongestFailureOrdering(AO);
+    auto Result = atomicInfo.EmitAtomicCompareExchangeLibcall(
+        AtomicLoadRes.second, NewAtomicAddr, AO, Failure);
+    LoadInst *PHILoad = Builder.CreateLoad(XElemTy, Result.first);
+    PHI->addIncoming(PHILoad, Builder.GetInsertBlock());
+    Builder.CreateCondBr(Result.second, ExitBB, ContBB);
+    OldVal->eraseFromParent();
+    Res.first = OldExprVal;
+    Res.second = Upd;
+
+    if (UnreachableInst *ExitTI =
+            dyn_cast<UnreachableInst>(ExitBB->getTerminator())) {
+      CurBBTI->eraseFromParent();
+      Builder.SetInsertPoint(ExitBB);
+    } else {
+      Builder.SetInsertPoint(ExitTI);
```
Could you add some comment explaining the code+control flow that is generated here?

Meinersbur reviewed (commented):
I see you squashed in my proposed refactoring of AtomicInfo. I think it should be its own PR. It also was just a sketch and needs to be completed (`check-clang` is failing) and needs some cleanup (`auto` -> type, `emitAtomicLibcalls2`). Ideally, it would be NFC for clang. I was hoping you would continue with my sketch, but I could work on it as well.

NimishMishra commented:
> I see you squashed in my proposed refactoring of AtomicInfo. I think it should be its own PR. It also was just a sketch and needs to be completed (`check-clang` is failing) and needs some cleanup (`auto` -> type, `emitAtomicLibcalls2`). Ideally, it would be NFC for clang. I was hoping you would continue with my sketch, but I could work on it as well.

I could complete the portion for Clang. I was currently working on supporting types (currently, the libcall arguments are accepted as pointers). Apart from cleanup and support for types, is there anything else you had in mind that I could cover?

NimishMishra commented:
RFC for refactoring CGAtomic: https://discourse.llvm.org/t/rfc-refactoring-cgatomic-into-llvmfrontend/80168

NimishMishra commented:
Hi all,

   Sorry for the long stall on this. Since the RFC https://discourse.llvm.org/t/rfc-refactoring-cgatomic-into-llvmfrontend/80168/8 for refactoring atomic libcall functionality to LLVM Frontend hit a roadblock, we thought it better to move ahead with the entire solution living in the IRBuilder. 

Does the current solution look good?

Meinersbur reviewed (approved):
LGTM

Meinersbur reviewed (commented):
@tblah Mentions that there is still a problem in that this emits libcall atomics (EmitAtomicLoadLibcall, `__atomic_compare_exchange`) that not all targets support. This was also what my CGAtomic refactoring proposal intended to solve.

NimishMishra commented:
> @tblah Mentions that there is still a problem in that this emits libcall atomics (EmitAtomicLoadLibcall, `__atomic_compare_exchange`) that not all targets support. This was also what my CGAtomic refactoring proposal intended to solve.


Thanks @Meinersbur for the review. Yes I agree. I was wondering whether the target's information can be exposed somehow in the IRBuilder or translation? If so, can we appropriately error out a TODO like "atomic update variable is a struct requiring an atomic libcall which is absent on the target". Alternatively, should we add a lightweight interface in LLVM Frontend that returns a decision whether the target supports the libcalls?

Meinersbur commented:
One of other proposals was to preliminarily add `emitAtomicCompareExchangeBuiltin` from https://github.com/llvm/llvm-project/pull/101966 for use in OpenMPIRBuilder only (potentially simplified by removing options that OpenMPIRBuilder does not need). It could be later moved to LLVMFrontendAtomic once there is agreement what the generalized interface should be.

Check for the existence of support of `__atomic_compare_exchange` exists in `BuildLibCall.cpp` by querying TargetLibraryInfo. But it is too late in the compilation stage to emit an error, so they only sensible thing to do at this point is calling `error:report_fatal_error` like AtomicExpandPass does (there is no legal fallback behaviour). Proper handling would be during semantic analysis and emitting diagnostic which would be a completely different code path. Clang's CGAtomic doesn't do so either and is happy to emit `__atomic_compare_exchange` regardless whether the target supports it or not, so I think it is acceptable to not do this in Flang either for now. The user will see a link error. It even made it into [some unittests](https://github.com/llvm/llvm-project/blob/3c83102f0615c7d66f6df698ca472ddbf0e9483d/llvm/test/Transforms/AtomicExpand/AMDGPU/expand-atomicrmw-fp-vector.ll#L25).

NimishMishra commented:
Okay thanks. We should be good to merge.

@kiranchandramohan @tblah Can we merge this?

kiranchandramohan reviewed (commented):
I think we need to organize and test this patch a bit better.

1. Is the following sentence in the summary still correct or is the code from LLVM AtomicExpand?
```
We first refactor relevant functionality from Clang Codegen to LLVM Frontend, and then interface the OpenMPIRBuilder with the same. 
```
2. The test removed from Todo is an atomic read, why was it not added back?
3. The code change in OpenMPToLLVMIRTranslation.cpp is the Capture Code generation. There is no test for this. Also cand this change be shared with the Update code generation?
4. We need tests in the OpenMP IRBuilder and in MLIR for the changes.

NimishMishra commented:
Thanks @kiranchandramohan. Yes, atomic capture can share this code; I'll add the tests.

NimishMishra commented:
Hi @kiranchandramohan, thanks for the review. 

> 
> 1. Is the following sentence in the summary still correct or is the code from LLVM AtomicExpand?

Updated the description.
 
> 2. The test removed from Todo is an atomic read, why was it not added back?

I thought it would be good to deal with it in another PR since it deals with atomic read. I have already created the PR in my local fork: https://github.com/NimishMishra/f18-llvm-project/commit/b61d1837b95e7f5dd431553f543f25e598100b0c; I will push it for review as soon as this PR lands (will add mlir tests before pushing).

We would also need a similar handling for atomic write, will push a PR for that too.

> 3. The code change in OpenMPToLLVMIRTranslation.cpp is the Capture Code generation. There is no test for this. Also cand this change be shared with the Update code generation?

   Added tests.

   Sharing code is possible, but it requires changing some interfaces. Reason being that `convertOmpAtomicCapture` directly calls `createAtomicCapture` in the IRBuilder, so it needs all details for the atomic update operation (which `convertOmpAtomicUpdate` cannot provide in its current function signature).

   One way could be to deprecate `createAtomicCapture` and emit the two operations directly in `OpenMPToLLVMIRTranslation.cpp`. Suppose we need to emit (update, read), then invoke `convertOmpAtomicUpdate` and `convertOmpAtomicRead`, which in turn invoke `createAtomicUpdate` and `createAtomicRead` in the IRBuilder respectively.

   Another way could be `convertOmpAtomicCapture` invokes `convertOmpAtomicUpdate` and passes a bunch of pointers such that `convertOmpAtomicUpdate` sends back all necessary details for the update operation. Then `convertOmpAtomicCapture` can invoke `createAtomicCapture` in the IRBuilder normally. This feels a bit cumbersome though.

> 4. We need tests in the OpenMP IRBuilder and in MLIR for the changes.

I've added some tests. Do they look okay?

NimishMishra commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7943,6 +7943,83 @@ Value *OpenMPIRBuilder::emitRMWOpAsInstruction(Value *Src1, Value *Src2,
+                                       llvm::AtomicOrdering AO,
+                                       uint64_t AtomicSizeInBits) {
+  LLVMContext &Ctx = Builder.getContext();
+  Type *SizedIntTy = Type::getIntNTy(Ctx, AtomicSizeInBits * 8);
```
Thanks for your review @tblah. 

@Meinersbur also suggested to let the atomic libcall codegen reside in the Frontend, since that is much cleaner. I have addressed your comments, but the code has shifted. I am detailing where the modified code changes in response to your review reside.

`Type *SizedIntTy = Type::getIntNTy(Ctx, getAtomicSizeInBits());` will indeed take the size in bits. This now resides in `EmitAtomicLoadLibcall` in `Atomic.h`

NimishMishra commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7943,6 +7943,83 @@ Value *OpenMPIRBuilder::emitRMWOpAsInstruction(Value *Src1, Value *Src2,
+  LLVMContext &Ctx = Builder.getContext();
+  uint64_t IntBits = 32;
+  uint16_t SizeTBits = 64;
+  uint16_t BitsPerByte = 8;
```
Changed these too to use `constexpr`. They now reside in `getAtomicSizeValue` in `Atomic.h`

NimishMishra commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7943,6 +7943,83 @@ Value *OpenMPIRBuilder::emitRMWOpAsInstruction(Value *Src1, Value *Src2,
+    llvm::Value *ExpectedVal, llvm::Value *DesiredVal,
+    llvm::AtomicOrdering Success, llvm::AtomicOrdering Failure) {
+  LLVMContext &Ctx = Builder.getContext();
+  uint64_t IntBits = 32;
```
The last two arguments to `__atomic_compare_exchange` are `int success` and `int failure`. Hence, we create a `llvm::Constant::getIntegerValue` and use the generic 4-byte width for an integer. I have added a TODO to instead get the width of an integer from `TargetMachine` and `TargetInfo`. The new code resides in `EmitAtomicCompareExchangeLibcall` in `Atomic.h`

NimishMishra commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7943,6 +7943,83 @@ Value *OpenMPIRBuilder::emitRMWOpAsInstruction(Value *Src1, Value *Src2,
+      DesiredVal,
+      llvm::Constant::getIntegerValue(
+          llvm::IntegerType::get(Ctx, IntBits),
+          llvm::APInt(IntBits, (uint64_t)Success, /*signed=*/true)),
```
Changed to `static_cast<uint64_t>(Success)` in `EmitAtomicCompareExchangeLibcall` in `Atomic.h`

NimishMishra commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7943,6 +7943,83 @@ Value *OpenMPIRBuilder::emitRMWOpAsInstruction(Value *Src1, Value *Src2,
+          llvm::APInt(IntBits, (uint64_t)Success, /*signed=*/true)),
+      llvm::Constant::getIntegerValue(
+          llvm::IntegerType::get(Ctx, IntBits),
+          llvm::APInt(IntBits, (uint64_t)Failure, /*signed=*/true)),
```
Changed to `static_cast<uint64_t>(Failure)` in `EmitAtomicCompareExchangeLibcall` in `Atomic.h`

NimishMishra commented:
The CI failures are from `llvm-libc++-shared.cfg`. I'll rebase the PR after a while

tblah commented on flang/lib/Lower/DirectivesCommon.h:
```diff
@@ -129,8 +129,7 @@ static void processOmpAtomicTODO(mlir::Type elementType,
     mlir::Type unwrappedEleTy = fir::unwrapRefType(elementType);
-    bool supportedAtomicType =
-        (!fir::isa_complex(unwrappedEleTy) && fir::isa_trivial(unwrappedEleTy));
+    bool supportedAtomicType = fir::isa_trivial(unwrappedEleTy);
```
I think we need more complex logic here. So far as I can tell this PR only fixes atomic update and capture. Using complex with atomic read (for example) still needs a TODO message.

tblah replied:
That's okay by me

NimishMishra committed:
[flang][mlir][llvm][OpenMP] Emit __atomic_load and __atomic_compare_exchange libcalls for complex types in atomic update

NimishMishra commented:
It seems that the aarch64 build bots are failing for atomic-capture integration test (the same IR is passing when the MLIR tests are running).

I'm not entirely sure where the issue could be. Should I submit another PR to remove this particular integration test (no changes will be needed to the MLIR tests) ?

Dinistro commented:
Some of the test failures look like LLVM going ahead and replacing something with a loop. This can happen when a specific intrinsic is not supported natively for a target architecture.

NimishMishra commented:
Merged https://github.com/llvm/llvm-project/pull/110969 that reverts the integration test `atomic-capture-complex.f90` and the following builds are okay now:

1. flang-aarch64-dylib
2. flang-aarch64-latest-gcc
3. flang-aarch64-libcxx
4. flang-aarch64-rel-assert
5. flang-aarch64-release
6. flang-aarch64-sharedlibs

tblah commented:
I think the problem might have been that `flang-new -fc1 -emit-llvm` will run an optimization pipeline on the LLVM IR before printing. Even without specifying `-O` I don't think the result will always be the same as taking the output of `mlir-translate`.

NimishMishra commented:
Oh. I didn't know that. Thanks.

DanielCChen commented:
@NimishMishra 
The test case `flang/test/Integration/OpenMP/atomic-capture-complex.f90` broke Flang buildbot on LoP https://lab.llvm.org/buildbot/#/builders/157

NimishMishra commented:
Hi @DanielCChen,

   I merged https://github.com/llvm/llvm-project/pull/110969 which removes this test case. Would a rebase and re-trigger help?

NimishMishra commented:
I see some recent failures on `flang/test/Integration/OpenMP/atomic-update-complex.f90` as now. I'm removing that as well.

DanielCChen commented:
Thanks @NimishMishra ! I will rebase and rebuild.

NimishMishra commented:
Thanks. Please rebase over https://github.com/llvm/llvm-project/commit/71b2c4dbc7cc28007849434303e6bd2408db5b3b. I think the build-bots would work fine now.

NimishMishra commented:
This build on my latest commit (https://lab.llvm.org/buildbot/#/builders/157/builds/9666) seems ok. Let me know if the issue is also solved on your end. Thanks!

NimishMishra commented on llvm/lib/Frontend/Atomic/Atomic.cpp:
```diff
@@ -0,0 +1,19 @@
+
+namespace {} // namespace
+
+namespace llvm {} // end namespace llvm
```
Thanks @nikic. I will prepare a NFC to address all of your comments.
